File: apps/ios/Pebbles/Features/Path/Models/EmotionCategoryOrdering.swift
import Foundation

/// Static curation data: the order in which the 7 emotion categories
/// surface in `EmotionPickerSheet`, indexed by the form's currently-selected
/// `Valence`.
///
/// The valence enum lives only in iOS, so this ordering deliberately lives
/// in iOS-only code rather than in the database. If a second client picks
/// up this UX later, the table-form is sketched in the spec.
///
/// Ordering rule (informally): own polarity first, opposite polarity last.
/// Within each polarity, the "peak" member leads at LARGE, the most
/// "subtle" member leads at SMALL, balanced at MEDIUM.
enum EmotionCategoryOrdering {

    struct Key: Hashable {
        let size: ValenceSizeGroup
        let polarity: ValencePolarity

        init(_ size: ValenceSizeGroup, _ polarity: ValencePolarity) {
            self.size = size
            self.polarity = polarity
        }
    }

    static let byValence: [Key: [String]] = [
        // HIGHLIGHTS — pleasant first
        Key(.large,  .highlight): ["pride",   "joy",     "peace",   "fear",    "anger",   "shame",   "sadness"],
        Key(.medium, .highlight): ["joy",     "pride",   "peace",   "fear",    "anger",   "shame",   "sadness"],
        Key(.small,  .highlight): ["peace",   "joy",     "pride",   "shame",   "sadness", "fear",    "anger"],

        // NEUTRALS — balanced, peace leads
        Key(.large,  .neutral):   ["peace",   "joy",     "pride",   "fear",    "anger",   "shame",   "sadness"],
        Key(.medium, .neutral):   ["peace",   "fear",    "joy",     "anger",   "pride",   "shame",   "sadness"],
        Key(.small,  .neutral):   ["peace",   "anger",   "joy",     "fear",    "pride",   "sadness", "shame"],

        // LOWLIGHTS — unpleasant first
        Key(.large,  .lowlight):  ["sadness", "fear",    "anger",   "shame",   "peace",   "joy",     "pride"],
        Key(.medium, .lowlight):  ["anger",   "fear",    "shame",   "sadness", "peace",   "pride",   "joy"],
        Key(.small,  .lowlight):  ["shame",   "sadness", "fear",    "anger",   "peace",   "pride",   "joy"],
    ]

    /// Used when no valence is selected on the draft yet. Equal to Medium Neutral.
    static let `default`: [String] = ["peace", "fear", "joy", "anger", "pride", "shame", "sadness"]

    static func order(for valence: Valence?) -> [String] {
        guard let v = valence else { return `default` }
        return byValence[Key(v.sizeGroup, v.polarity)] ?? `default`
    }
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/features/path/models/EmotionCategoryOrdering.kt
package app.pbbls.android.features.path.models

object EmotionCategoryOrdering {
    val byValence: Map<Pair<ValenceSizeGroup, ValencePolarity>, List<String>> =
        mapOf(
            Pair(ValenceSizeGroup.LARGE, ValencePolarity.HIGHLIGHT) to
                listOf("pride", "joy", "peace", "fear", "anger", "shame", "sadness"),
            Pair(ValenceSizeGroup.MEDIUM, ValencePolarity.HIGHLIGHT) to
                listOf("joy", "pride", "peace", "fear", "anger", "shame", "sadness"),
            Pair(ValenceSizeGroup.SMALL, ValencePolarity.HIGHLIGHT) to
                listOf("peace", "joy", "pride", "shame", "sadness", "fear", "anger"),
            Pair(ValenceSizeGroup.LARGE, ValencePolarity.NEUTRAL) to
                listOf("peace", "joy", "pride", "fear", "anger", "shame", "sadness"),
            Pair(ValenceSizeGroup.MEDIUM, ValencePolarity.NEUTRAL) to
                listOf("peace", "fear", "joy", "anger", "pride", "shame", "sadness"),
            Pair(ValenceSizeGroup.SMALL, ValencePolarity.NEUTRAL) to
                listOf("peace", "anger", "joy", "fear", "pride", "sadness", "shame"),
            Pair(ValenceSizeGroup.LARGE, ValencePolarity.LOWLIGHT) to
                listOf("sadness", "fear", "anger", "shame", "peace", "joy", "pride"),
            Pair(ValenceSizeGroup.MEDIUM, ValencePolarity.LOWLIGHT) to
                listOf("anger", "fear", "shame", "sadness", "peace", "pride", "joy"),
            Pair(ValenceSizeGroup.SMALL, ValencePolarity.LOWLIGHT) to
                listOf("shame", "sadness", "fear", "anger", "peace", "pride", "joy"),
        )

    /** Used when no valence is selected yet. Equal to MEDIUM · NEUTRAL. */
    val default: List<String> = listOf("peace", "fear", "joy", "anger", "pride", "shame", "sadness")

    fun order(valence: Valence?): List<String> {
        valence ?: return default
        return byValence[Pair(valence.sizeGroup, valence.polarity)] ?: default
    }
}
